src/Relations/BelongsToMany.php

```diff
@@ -257,7 +259,7 @@ protected function buildDictionary(Collection $results)
 
         foreach ($results as $result) {
             foreach ($result->$foreign as $item) {
-                $dictionary[$item][] = $result;
+                $dictionary[(string) $item][] = $result;
```

Comment:
Casts the `ObjectId` to a valid array-key.

The type of `$dictionary` can't be changed to `SplObjectStorage` it's used to search on values in Laravel's [`BelongsToMany::match`](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php#L287).

Fortunately, the key is [stringified](https://github.com/laravel/framework/blob/81ae53fcdb1a552237ef73f65f1f3faac8236641/src/Illuminate/Database/Eloquent/Relations/Concerns/InteractsWithDictionary.php#L19) before being searched.